File: 2021/src/solutions/day_01.rs
use aoc_common::{load, Solution};

pub struct Day01 {}

impl Solution for Day01 {

    fn name(&self) -> String {
        "Sonar Sweep".to_owned()
    }

    fn part_1(&self) -> String {
        let measurements = load("day_01")
            .lines()
            .map(|x| x.parse::<u32>().unwrap())
            .collect::<Vec<u32>>();

        let increases = measurements
            .windows(2)
            .filter(|x| x[0] < x[1])
            .count();

        increases.to_string()
    }

    fn part_2(&self) -> String {
        let measurements = load("day_01")
            .lines()
            .map(|x| x.parse::<u32>().unwrap())
            .collect::<Vec<u32>>();

        let mut increases = 0;
        for i in 3..measurements.len() {
            let a = measurements[i - 1] + measurements[i - 2] + measurements[i - 3];
            let b = measurements[i] + measurements[i - 1] + measurements[i - 2];

            if b > a { increases += 1; }
        }

        increases.to_string()
    }
}

File: src/AdventOfCode.AoC2024/Solution/Day01.cs
using AdventOfCode.Core;

namespace AdventOfCode.AoC2024.Solution;

public class Day01 : ISolution
{
    public string Name => "Day 1";

    public string Part1(string inputFile)
    {
        var data = Calendar.LoadInput(inputFile);
        var (left, right) = ParseInput(data);
        
        left.Sort();
        right.Sort();
        
        var output = left.Zip(right, (l, r) => Math.Abs(l - r));
        return output.Sum().ToString();
    }

    public string Part2(string inputFile)
    {
        var data = Calendar.LoadInput(inputFile);
        var (left, right) = ParseInput(data);
        
        var output = left.Select(l => l * right.Count(r => l == r))
            .ToArray();
        
        return output.Sum().ToString();
    }

    private static (List<int> left, List<int> right) ParseInput(string input)
    {
        var lines = input.Split(Environment.NewLine);
        var left = new List<int>(lines.Length);
        var right = new List<int>(lines.Length);

        foreach (var line in lines)
        {
            var split = line.Split("   ");
            left.Add(int.Parse(split[0].Trim()));
            right.Add(int.Parse(split[1].Trim()));
        }
        
        return (left, right);
    }
}